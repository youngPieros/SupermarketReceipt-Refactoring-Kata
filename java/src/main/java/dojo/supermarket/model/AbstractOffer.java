package dojo.supermarket.model;

public abstract class AbstractOffer implements Offer {
    SpecialOfferType offerType;
}
