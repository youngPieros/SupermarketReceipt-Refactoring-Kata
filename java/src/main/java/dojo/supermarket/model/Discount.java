package dojo.supermarket.model;

public class Discount {
    private final String description;
    private final double discountAmount;
    private final Product product;

    private Discount(Product product, String description, double discountAmount) {
        this.description = description;
        this.discountAmount = discountAmount;
        this.product = product;
    }

    public static Discount createDiscount(Product product, String description, double discountAmount) {
        return new Discount(product, description, discountAmount);
    }

    public String getDescription() {
        return description;
    }

    public double getDiscountAmount() {
        return discountAmount;
    }

    public Product getProduct() {
        return product;
    }
}
