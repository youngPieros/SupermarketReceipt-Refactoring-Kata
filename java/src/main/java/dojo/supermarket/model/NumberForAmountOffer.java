package dojo.supermarket.model;

import java.util.Map;

public class NumberForAmountOffer extends SingleProductOffer {

    private final int numberOfAmountOffer;
    private final double amount;

    public NumberForAmountOffer(Product product, int numberOfAmountOffer, double amount) {
        super(product);
        this.numberOfAmountOffer = numberOfAmountOffer;
        this.amount = amount;
    }

    @Override
    public Discount calculateDiscount(Map<Product, Double> productQuantities) {
        int quantity = productQuantities.get(this.getProduct()).intValue();
        int quantityNumberWithDiscount = (quantity / numberOfAmountOffer * numberOfAmountOffer);
        double discount = quantityNumberWithDiscount * (this.getProduct().getPrice() - this.amount / numberOfAmountOffer);
        return Discount.createDiscount(this.getProduct(), this.numberOfAmountOffer + " for " + this.amount, discount);
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return productQuantities.get(getProduct()) >= numberOfAmountOffer;
    }

}
