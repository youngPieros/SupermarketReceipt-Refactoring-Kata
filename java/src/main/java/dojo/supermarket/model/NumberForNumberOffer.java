package dojo.supermarket.model;

import java.util.Map;

public class NumberForNumberOffer extends SingleProductOffer {

    private final int numberOfBoughtProducts;
    private final int numberOfPaidProducts;

    public NumberForNumberOffer(Product product, int numberOfPaidProducts, int numberOfBoughtProducts) {
        super(product);
        this.numberOfBoughtProducts = numberOfBoughtProducts;
        this.numberOfPaidProducts = numberOfPaidProducts;
    }

    @Override
    protected Discount calculateDiscount(Map<Product, Double> productQuantities) {
        int quantity = productQuantities.get(this.getProduct()).intValue();
        int quantityNumberWithDiscount = quantity / numberOfBoughtProducts;
        double discount = quantityNumberWithDiscount * (this.numberOfBoughtProducts - this.numberOfPaidProducts) * this.getProduct().getPrice();
        String description = numberOfBoughtProducts + " for " + numberOfPaidProducts;
        return Discount.createDiscount(this.getProduct(), description, discount);
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return productQuantities.get(getProduct()) >= numberOfBoughtProducts;
    }
}
