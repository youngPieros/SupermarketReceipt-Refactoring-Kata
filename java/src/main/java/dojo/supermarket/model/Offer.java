package dojo.supermarket.model;
import java.util.*;

public interface Offer {
    Discount calculateDiscount(Map<Product, Double> productQuantities);

    Double getIncludedProductDiscountNumber(Map<Product, Double> productQuantities);
}
