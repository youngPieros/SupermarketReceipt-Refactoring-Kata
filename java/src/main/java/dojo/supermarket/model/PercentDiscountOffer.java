package dojo.supermarket.model;

import java.util.Map;

public class PercentDiscountOffer extends AbstractOffer {

    private final Product product;
    private final double amountPercent;

    public PercentDiscountOffer(Product product, double amountPercent) {
        this.product = product;
        this.amountPercent = amountPercent;
    }


    @Override
    public Discount calculateDiscount(Map<Product, Double> productQuantities) {
        double quantityNumberWithDiscount = getIncludedProductDiscountNumber(productQuantities);
        double discount = quantityNumberWithDiscount * product.getPrice() * amountPercent / 100;
        String description = this.amountPercent + "% off";
        return Discount.createDiscount(description, discount);
    }

    @Override
    public Double getIncludedProductDiscountNumber(Map<Product, Double> productQuantities) {
        return productQuantities.get(product);
    }
}
