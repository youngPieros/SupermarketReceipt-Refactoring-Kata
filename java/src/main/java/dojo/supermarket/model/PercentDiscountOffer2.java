package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class PercentDiscountOffer2 extends BundleProductOffer {

    public PercentDiscountOffer2(ArrayList<Product> products) {
        super(products);
    }

    @Override
    public List<Discount> calculateDiscounts(Map<Product, Double> productQuantities) {
        return null;
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return true;
    }
}
