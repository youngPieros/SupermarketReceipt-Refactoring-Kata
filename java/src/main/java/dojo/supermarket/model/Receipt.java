package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.List;

public class Receipt {
    private List<ReceiptItem> items = new ArrayList<>();

    public Double getTotalPrice() {
        double total = 0.0;
        for (ReceiptItem item : this.items) {
            total += item.getTotalPrice();
        }
        return total;
    }

    public void addProduct(ProductQuantity productQuantity) {
        for (ReceiptItem receiptItem: this.items)
            if (receiptItem.getProduct().equals(productQuantity.getProduct())) {
                receiptItem.addQuantity(productQuantity.getQuantity());
                return;
            }
        this.items.add(ReceiptItem.createFromProductQuantity(productQuantity));
    }

    public List<ReceiptItem> getItems() {
        return new ArrayList<>(this.items);
    }

    public void addDiscount(Product product, Discount discount) {
        for (ReceiptItem receiptItem: items) {
            if (receiptItem.getProduct().equals(product)) {
                receiptItem.setDiscount(discount);
            }
        }
    }

}
