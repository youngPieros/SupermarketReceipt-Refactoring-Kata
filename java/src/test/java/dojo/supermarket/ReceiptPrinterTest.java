package dojo.supermarket;

import dojo.supermarket.model.*;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

public class ReceiptPrinterTest {

    Product toothbrush = new Product("toothbrush", 0.99, ProductUnit.Each);
    Product apples = new Product("apples", 1.99, ProductUnit.Kilo);
    Receipt receipt = new Receipt();

    @Test
    public void oneLineItem() {
        receipt.addProduct(new ProductQuantity(toothbrush, 1));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void quantityTwo() {
        receipt.addProduct(new ProductQuantity(toothbrush, 2));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void looseWeight() {
        receipt.addProduct(new ProductQuantity(apples, 2.3));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void total() {
        receipt.addProduct(new ProductQuantity(toothbrush, 1));
        receipt.addProduct(new ProductQuantity(apples, 0.75));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void discounts() {
        receipt.addDiscount(new Discount(apples, "3 for 2", -0.99));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void printWholeReceipt() {
        receipt.addProduct(new ProductQuantity(toothbrush, 1));
        receipt.addProduct(new ProductQuantity(toothbrush, 2));
        receipt.addProduct(new ProductQuantity(apples, 0.75));
        receipt.addDiscount(new Discount(toothbrush, "3 for 2", -0.99));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

}
