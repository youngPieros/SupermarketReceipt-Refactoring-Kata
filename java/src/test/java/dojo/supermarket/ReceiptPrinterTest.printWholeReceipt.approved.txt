toothbrush                          2.97
  0.99 * 3
apples                              1.49
  1.99 * 0.750
3 for 2(toothbrush)                -0.99

Total:                              3.47
